### PR TITLE
RFC: Sub-Commands

### PR DESCRIPTION
**Pull request by jwbixby**
Issue: https://github.com/encode/apistar/issues/236

## Introduction

This RFC goes over a potential implementation of Sub-Commands. The goal of the Sub-Commands feature is to simply the command line interface by allowing commands to be grouped and documented.

The following section merely introduces the implementation. Readers should reference the updated `README.md` documentation for examples on how to apply these features.

## Implementation

This RFC introduces two main pieces to achieve the features goals.

1. `SubCommand` - The ability to group commands together under a parent command.
2. Documentation - The ability to add inline documentation that enhances the the command line interface with more information and details.

### SubCommand

The `SubCommand` class is similar to the `Environment` class in the sense that it allows users to define class properties that will be used later when the user instantiates it. This allows the user to define properties that should work in most cases, but later these values can be overridden.

One situation where a user might want to override these values is when this SubCommand class is defined by a third party(e.g. apistar-sqlalchemy). Users can easily override the predefined `name` property to any name they'd like.

### Documenation

I believe this section is best explained by the documentation added to README.md. TLDR; By standardizing how documentation is written in the doc string, we can provide a better interface in the command line.

---

After preparing this, I did some thinking on how this could be applied to Routes. I left some notes on a branch in my fork that shows how these could be applied in a very similar way and help document nested routes. I'll leave a link here, but its out of scope for this PR. https://github.com/jwbixby/apistar/blob/jb_future_of_routes/route_explaination.md

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -294,7 +297,8 @@ app = App(routes=[
 
 ### Reversing URLS
 
-Rather than build URLs by hand, it is possible to generate URLs based on the endpoint, using `reverse_url()`.
+Rather than build URLs by hand, it is possible to generate URLs based on the
+endpoint, using `reverse_url()`.
```
Comment (audiolion) [inline README.md]: In markdown, you shouldn't insert your own linebreaks, leave paragraphs all on the same line

File: README.md [hunk 2]
```diff
@@ -320,10 +324,19 @@ app = App(routes=[
 ])
 ```
 
-## Command Routing
+---
+
+# Commands
+
+In addition to routing URLs, you can also route commands, to make
+additional functionality available directly to the command line client.
```
Comment (audiolion) [inline README.md]: remove linebreak

File: README.md [hunk 3]
```diff
@@ -345,6 +358,126 @@ commands = [
...
+* `name` - A string that will identify the group of commands. If not
+specified, the `SubCommand` will automatically generate a name based on the class
+name by truncating `"Commands"`. For example, `UserCommands` will be under the
+command named `user`.
+* `commands` - A list of `Command`s or `SubCommand`s. These will be the nested commands.
+
+### Example
+
+The following `app` configuration will have 4 top level commands: `user`,
+`override_user`, `import`, and `other_command`.
```
Comment (audiolion) [inline README.md]: Remove linebreak

File: README.md [hunk 4]
```diff
@@ -345,6 +358,126 @@ commands = [
...
+```
+
+### Documenting Commands
+
+Commands are documented by the doc string on the extended `SubCommand` class and
+command functions. Both follow the same format of a short description on the first
+line, an empty line, and a longer description starting on the third line. The
+longer description can be any length. Failure to separate the short and long
+description with an empty line will result in only the long description being
+specified.
```
Comment (audiolion) [inline README.md]: remove above linebreaks

**Comment by jwbixby** [comment]
I love the new `Document` idea for `0.4.0`. Keep up the great work! Can't wait for this framework to be stable.

**Pull request closed without merging**